juju/juju
# chore: cleanup legacy state charm methods

This patch removes most of the legacy Charm methods, as well as private methods and structs used by these methods.
This can be done without major issues because we were not double writing anymore at this point.

The only public method that is left for future cleanup is `LatestPlaceholderCharm()` which is still used in the status (although no data is returned), so it can be reimplemented correctly using the charm domain in a future patch.


## QA steps

This patch should have no impacts on the usual workflows since we weren't double writing anymore, so bootstrapping, deploying and migrating should still work: 


 1. Bootstrap a 3.6 controller and deploy a charm.

```sh
$ juju bootstrap lxd src36
$ juju add-model moveme1
$ juju deploy juju-qa-test
```

 2. Bootstrap a 4.0 controller with the changes and migrate the model.

```sh
$ juju bootstrap lxd dst40
$ juju migrate src36:moveme1 dst40
$ juju add-unit juju-qa-test
```

 3. Verify no errors exist in the model logs for the agents. If there are
    errors, this is a bug and should be fixed before merging. The fix can
    either be applied to the 4.0 branch (preferable) or the 3.6 branch, though
    that needs to be discussed with the team.

```sh
$ juju debug-log -m dst40:controller
$ juju debug-log -m dst40:moveme1
```

4. We also need to test a model migration from 4.0 to 4.0.

```sh
$ juju bootstrap lxd src40
$ juju add-model moveme2
$ juju deploy juju-qa-test
```

```sh
$ juju migrate src40:moveme2 dst40
$ juju add-unit juju-qa-test
```

5. Verify that there are no errors in the controller or model logs.

```sh
$ juju debug-log -m dst40:controller
$ juju debug-log -m dst40:moveme2
```

## Documentation changes



## Links



**Launchpad bug:** https://bugs.launchpad.net/juju/+bug/

**Jira card:** JUJU-
